File: python/techai_webutils/src/techai_webutils/execution/job/decorators/builder.py
"""Job decorators: logging/metrics/tracing/recovery AnyJob wrappers + a Wrap builder.

The async port of Go ``job/decorators`` — compose via
``wrap(job).with_logger(..).with_metrics(..).with_tracing(..).with_recovery().build()``. Decorators
layer innermost->outermost as recovery -> logging -> metrics -> tracing, so a raised exception is
converted to a failure first and the outer layers observe it as a normal outcome. A chain with no
decorators returns the inner job unchanged.
"""

from __future__ import annotations

import asyncio
from typing import TYPE_CHECKING, Self

from techai_webutils.core.interfaces.execution import BatchResult, StepResult, StepStatus

if TYPE_CHECKING:
    from techai_webutils.core.interfaces.execution import AnyJob, JobMeta
    from techai_webutils.core.interfaces.logger import Logger
    from techai_webutils.core.interfaces.metrics import MetricCounter, MetricHistogram, MetricsProvider
    from techai_webutils.core.interfaces.tracer import TracerProvider


class _LoggingJob:
    """Wraps an AnyJob with structured start/finish/error logging (mirrors Go ``loggingJob``)."""

    def __init__(self, inner: AnyJob, logger: Logger) -> None:
        self._inner: AnyJob = inner
        self._logger: Logger = logger

    def meta(self) -> JobMeta:
        """Return the wrapped job's metadata unchanged."""
        return self._inner.meta()

    async def execute(self) -> BatchResult:
        """Log the job start, run it, then log the outcome + elapsed (error log + re-raise on failure)."""
        meta = self._inner.meta()
        self._logger.info("job started", name=meta.name, group=meta.group)
        loop = asyncio.get_running_loop()
        start = loop.time()
        try:
            result = await self._inner.execute()
        except Exception as exc:
            self._logger.exception("job error", name=meta.name, elapsed=loop.time() - start, error=str(exc))
            raise
        elapsed = loop.time() - start
        if result.has_failures:
            self._logger.warning("job finished with failures", name=meta.name, elapsed=elapsed)
        else:
            self._logger.info("job finished", name=meta.name, elapsed=elapsed)
        return result


class _MetricsJob:
    """Wraps an AnyJob, recording an outcome-labeled run counter + duration (mirrors Go ``metricsJob``)."""

    def __init__(self, inner: AnyJob, metrics: MetricsProvider) -> None:
        self._inner: AnyJob = inner
        self._runs: MetricCounter = metrics.counter(
            "execution_job_runs_total",
            "Job runs by name and outcome",
            ["job", "outcome"],
        )
        self._duration: MetricHistogram = metrics.histogram(
            "execution_job_duration_seconds",
            "Job duration in seconds",
            labels=["job"],
        )

    def meta(self) -> JobMeta:
        """Return the wrapped job's metadata unchanged."""
        return self._inner.meta()

    async def execute(self) -> BatchResult:
        """Run the job, recording its duration and an outcome-labeled run counter.

        The outcome is ``ok`` / ``failed`` (from the result) or ``error`` when the inner job *raises*.
        Note ``error`` is only reachable when recovery is NOT wired: the standard
        ``recovery -> logging -> metrics -> tracing`` stack converts a raised exception into a FAIL
        result innermost, so metrics then observes ``failed`` (not ``error``).
        """
        meta = self._inner.meta()
        loop = asyncio.get_running_loop()
        start = loop.time()
        try:
            result = await self._inner.execute()
        except Exception:
            self._record(meta.name, "error", loop.time() - start)
            raise
        self._record(meta.name, "failed" if result.has_failures else "ok", loop.time() - start)
        return result

    def _record(self, job: str, outcome: str, elapsed: float) -> None:
        """Emit the run counter (by outcome) and the duration histogram."""
        self._runs.inc(1.0, job=job, outcome=outcome)
        self._duration.observe(elapsed, job=job)


class _TracingJob:
    """Wraps an AnyJob in a per-run span, recording errors on it (mirrors Go ``tracingJob``)."""

    def __init__(self, inner: AnyJob, tracer: TracerProvider) -> None:
        self._inner: AnyJob = inner
        self._tracer: TracerProvider = tracer

    def meta(self) -> JobMeta:
        """Return the wrapped job's metadata unchanged."""
        return self._inner.meta()

    async def execute(self) -> BatchResult:
        """Run the inner job within a span named for the job, recording a raised error on it."""
        meta = self._inner.meta()
        with self._tracer.span(f"job.{meta.name}") as span:
            try:
                return await self._inner.execute()
            except Exception as exc:
                span.record_error(exc)
                raise


class _RecoveryJob:
    """Wraps an AnyJob, converting a raised exception into a FAIL result (mirrors Go ``recoveryJob``)."""

    def __init__(self, inner: AnyJob, logger: Logger | None = None) -> None:
        self._inner: AnyJob = inner
        self._logger: Logger | None = logger

    def meta(self) -> JobMeta:
        """Return the wrapped job's metadata unchanged."""
        return self._inner.meta()

    async def execute(self) -> BatchResult:
        """Run the inner job, converting a non-cancellation exception into a single FAIL result."""
        meta = self._inner.meta()
        try:
            return await self._inner.execute()
        except asyncio.CancelledError:
            raise  # never swallow cancellation
        except Exception as exc:
            if self._logger is not None:
                self._logger.exception("job exception recovered", name=meta.name, error=str(exc))
            return BatchResult(
                results=(
                    StepResult(name=meta.name, group=meta.group, status=StepStatus.FAIL, error=str(exc)),
                ),
            )


class DecoratorBuilder:
    """Fluent builder wrapping an AnyJob with optional logging/metrics/tracing/recovery decorators."""

    def __init__(self, inner: AnyJob) -> None:
        """Start a decoration chain for ``inner``."""
        self._inner: AnyJob = inner
        self._logger: Logger | None = None
        self._metrics: MetricsProvider | None = None
        self._tracer: TracerProvider | None = None
        self._recover: bool = False

    def with_logger(self, logger: Logger) -> Self:
        """Add structured start/finish/error logging around the job."""
        self._logger = logger
        return self

    def with_metrics(self, metrics: MetricsProvider) -> Self:
        """Add run-count + duration metrics around the job."""
        self._metrics = metrics
        return self

    def with_tracing(self, tracer: TracerProvider) -> Self:
        """Add a per-run span around the job."""
        self._tracer = tracer
        return self

    def with_recovery(self) -> Self:
        """Add exception -> FAIL recovery (innermost, so logging/metrics/tracing observe the failure)."""
        self._recover = True
        return self

    def build(self) -> AnyJob:
        """Return the decorated job, layered recovery -> logging -> metrics -> tracing.

        A chain with no decorators returns the inner job unchanged.
        """
        result: AnyJob = self._inner
        if self._recover:
            result = _RecoveryJob(result, self._logger)
        if self._logger is not None:
            result = _LoggingJob(result, self._logger)
        if self._metrics is not None:
            result = _MetricsJob(result, self._metrics)
        if self._tracer is not None:
            result = _TracingJob(result, self._tracer)
        return result


def wrap(job: AnyJob) -> DecoratorBuilder:
    """Start a decoration chain for ``job`` (mirrors Go ``decorators.Wrap``)."""
    return DecoratorBuilder(job)

File: go/execution/job/decorators/builder.go
// Package decorators provides logging, metrics, tracing, and recovery wrappers
// for interfaces.AnyJob. They compose via Wrap(job).WithLogger(..).WithMetrics(..).
// WithTracing(..).WithRecovery().Build(), and are backward-compatible: a chain with
// no decorators returns the inner job unchanged.
package decorators

import (
	"context"
	"fmt"
	"time"

	coreerr "github.com/gt-tech-ai/knowledge-engine/go/core/errors"
	"github.com/gt-tech-ai/knowledge-engine/go/core/interfaces"
	"github.com/gt-tech-ai/knowledge-engine/go/core/types"
)

// loggingJob wraps an AnyJob with structured logging on start/finish.
type loggingJob struct {
	// inner is the wrapped job whose execution is logged.
	inner interfaces.AnyJob

	// logger receives the start, finish, and error log lines.
	logger interfaces.Logger
}

// Compile-time assertion that *loggingJob satisfies interfaces.AnyJob.
var _ interfaces.AnyJob = (*loggingJob)(nil)

// Meta returns the wrapped job's metadata unchanged.
func (l *loggingJob) Meta() types.JobMeta { return l.inner.Meta() }

// Execute logs the job start, runs the inner job, then logs its outcome
// (error or finished-with-failure-flag) along with the elapsed wall-clock.
func (l *loggingJob) Execute(
	ctx context.Context,
	runner interfaces.CommandRunner,
	root string,
) (types.StepResults, error) {
	meta := l.inner.Meta()
	l.logger.WithContext(ctx).Info("job started", "name", meta.Name, "group", meta.Group)
	start := time.Now()
	results, err := l.inner.Execute(ctx, runner, root)
	elapsed := time.Since(start)
	if err != nil {
		l.logger.WithContext(ctx).
			Error("job error", "name", meta.Name, "elapsed", elapsed, "err", err)
	} else {
		l.logger.WithContext(ctx).Info(
			"job finished",
			"name",
			meta.Name,
			"elapsed",
			elapsed,
			"failures",
			results.HasFailures(),
		)
	}
	return results, err
}

// metricsJob wraps an AnyJob, recording run count (labeled by outcome) and
// duration, so every run of a periodic job emits run/error counters and a
// latency histogram.
type metricsJob struct {
	// inner is the wrapped job whose runs are measured.
	inner interfaces.AnyJob
	// metrics is the registry the counter/histogram are created on.
	metrics interfaces.Metrics
}

// Compile-time assertion that *metricsJob satisfies interfaces.AnyJob.
var _ interfaces.AnyJob = (*metricsJob)(nil)

// Meta returns the wrapped job's metadata unchanged.
func (m *metricsJob) Meta() types.JobMeta { return m.inner.Meta() }

// Execute records the job's duration and an outcome-labeled run counter.
func (m *metricsJob) Execute(
	ctx context.Context,
	runner interfaces.CommandRunner,
	root string,
) (types.StepResults, error) {
	meta := m.inner.Meta()
	start := time.Now()
	results, err := m.inner.Execute(ctx, runner, root)
	elapsed := time.Since(start)

	outcome := "ok"
	switch {
	case err != nil:
		outcome = "error"
	case results.HasFailures():
		outcome = "failed"
	}
	m.metrics.Counter("execution_job_runs_total", "Job runs by name and outcome", "job", "outcome").
		Add(1, meta.Name, outcome)
	m.metrics.Histogram("execution_job_duration_seconds", "Job duration in seconds", nil, "job").
		Observe(elapsed.Seconds(), meta.Name)
	return results, err
}

// tracingJob wraps an AnyJob in a span, recording errors on it.
type tracingJob struct {
	// inner is the wrapped job whose execution is traced.
	inner interfaces.AnyJob
	// tracer starts the per-run span.
	tracer interfaces.Tracer
}

// Compile-time assertion that *tracingJob satisfies interfaces.AnyJob.
var _ interfaces.AnyJob = (*tracingJob)(nil)

// Meta returns the wrapped job's metadata unchanged.
func (t *tracingJob) Meta() types.JobMeta { return t.inner.Meta() }

// Execute runs the inner job within a span named for the job.
func (t *tracingJob) Execute(
	ctx context.Context,
	runner interfaces.CommandRunner,
	root string,
) (types.StepResults, error) {
	meta := t.inner.Meta()
	ctx, span := t.tracer.Start(ctx, "job."+meta.Name)
	defer span.End()
	results, err := t.inner.Execute(ctx, runner, root)
	if err != nil {
		span.RecordError(err)
	}
	return results, err
}

// recoveryJob wraps an AnyJob, converting a panic into an error + a failed
// StepResults so one panicking sweep run never crashes the long-running worker.
type recoveryJob struct {
	// inner is the wrapped job whose panics are recovered.
	inner interfaces.AnyJob
	// logger, when set, records the recovered panic (may be nil).
	logger interfaces.Logger
}

// Compile-time assertion that *recoveryJob satisfies interfaces.AnyJob.
var _ interfaces.AnyJob = (*recoveryJob)(nil)

// Meta returns the wrapped job's metadata unchanged.
func (r *recoveryJob) Meta() types.JobMeta { return r.inner.Meta() }

// Execute runs the inner job, recovering any panic into a returned error.
func (r *recoveryJob) Execute(
	ctx context.Context,
	runner interfaces.CommandRunner,
	root string,
) (results types.StepResults, err error) {
	meta := r.inner.Meta()
	defer func() {
		if rec := recover(); rec != nil {
			if r.logger != nil {
				r.logger.WithContext(ctx).
					Error("job panic recovered", "name", meta.Name, "panic", rec)
			}
			err = coreerr.New(
				coreerr.CodeInternal,
				fmt.Sprintf("job %q panicked: %v", meta.Name, rec),
			)
			results = types.StepResults{{
				Name:   meta.Name,
				Group:  meta.Group,
				Status: types.StatusFail,
				Error:  err.Error(),
			}}
		}
	}()
	return r.inner.Execute(ctx, runner, root)
}

// Builder wraps an AnyJob with optional logging, metrics, tracing, and recovery
// decorators.
type Builder struct {
	// inner is the base job being decorated.
	inner interfaces.AnyJob

	// logger, when set via WithLogger, adds the logging decorator on Build.
	logger interfaces.Logger

	// metrics, when set via WithMetrics, adds the metrics decorator on Build.
	metrics interfaces.Metrics

	// tracer, when set via WithTracing, adds the tracing decorator on Build.
	tracer interfaces.Tracer

	// recover, when set via WithRecovery, adds the panic-recovery decorator on Build.
	recover bool
}

// Wrap starts a decoration chain for job.
func Wrap(job interfaces.AnyJob) *Builder { return &Builder{inner: job} }

// WithLogger adds structured logging around the job.
func (b *Builder) WithLogger(logger interfaces.Logger) *Builder {
	b.logger = logger
	return b
}

// WithMetrics adds run-count + duration metrics around the job.
func (b *Builder) WithMetrics(metrics interfaces.Metrics) *Builder {
	b.metrics = metrics
	return b
}

// WithTracing adds a per-run span around the job.
func (b *Builder) WithTracing(tracer interfaces.Tracer) *Builder {
	b.tracer = tracer
	return b
}

// WithRecovery adds panic→error recovery (innermost, so logging/metrics/tracing
// still observe the converted failure).
func (b *Builder) WithRecovery() *Builder {
	b.recover = true
	return b
}

// Build returns the decorated job. Decorators are layered innermost→outermost as
// recovery → logging → metrics → tracing, so a panic is converted first and the
// outer layers observe it as a normal failure. A chain with no decorators returns
// the inner job unchanged.
func (b *Builder) Build() interfaces.AnyJob {
	result := b.inner
	if b.recover {
		result = &recoveryJob{inner: result, logger: b.logger}
	}
	if b.logger != nil {
		result = &loggingJob{inner: result, logger: b.logger}
	}
	if b.metrics != nil {
		result = &metricsJob{inner: result, metrics: b.metrics}
	}
	if b.tracer != nil {
		result = &tracingJob{inner: result, tracer: b.tracer}
	}
	return result
}
